dagshub/common/api/repo.py

```diff
@@ -226,6 +231,83 @@ def get_storage_file(self, path: str) -> bytes:
             raise RuntimeError(error_msg)
         return res.content
 
+    def _get_files_in_path(self, path, revision=None, recursive=False) -> List[ContentAPIEntry]:
+        """
+        Walks through the path of the repo, returning non-dir entries
+        """
+
+        dir_queue = [path]
+        files = []
+
+        progress = get_rich_progress(rich.progress.MofNCompleteColumn())
+        task = progress.add_task("Traversing directories...", total=None)
+
+        def step(step_path):
+            res = self.list_path(step_path, revision)
+            for entry in res:
+                if entry.type == "file":
+                    files.append(entry)
+                elif entry.type == "dir" and recursive:
+                    dir_queue.append(entry.path)
+                # TODO: storage
+            progress.update(task, advance=1)
+
+        with progress:
+            while len(dir_queue):
+                query_path = dir_queue.pop(0)
+                step(query_path)
+
+        return files
+
+    def download(
+        self, remote_path, local_path=".", revision=None, recursive=True, keep_source_prefix=True, redownload=False
+    ):
+        """
+        Downloads the contents of the repository at "remote_path" to the "local_path"
+
+        Args:
+            remote_path: path of the folder or folder to download in the repository
+            local_path: where to download the files. Defaults to CWD
+            revision: repo revision, if not specified - uses default repo branch
+            recursive: whether to download files recursively
+            keep_source_prefix: whether to keep the path of the folder in the download path or not
+                example: Given remote_path "src/data" and file "test/file.txt"
+                if True: will download to "<local_path>/src/data/test/file.txt"
+                if False: will download to "<local_path>/test/file.txt"
+            redownload: whether to redownload an already existing file
+                The downloader doesn't do any hash comparisons and only checks if a file already exists in the FS or not
+        """
+        files = self._get_files_in_path(remote_path, revision, recursive)
+        file_tuples = []
+        if local_path is None:
+            local_path = "."
+        local_path = Path(local_path)
+        # Strip the slashes from the beginning so the relative_to logic works
+        if str(remote_path.startswith("/")):
+            remote_path = str(remote_path).lstrip("/")
+        # Edge case - if the user requested a single file - different output path semantics
+        if len(files) == 1 and files[0].path == remote_path:
+            f = files[0]
+            remote_path = PurePosixPath(f.path)
+            # If local_path was specified, assume that the local_path is the exact name of the file
+            if local_path != Path("."):
+                file_path = local_path
```

Comment:
The local path might also be a directory, and to maintain the same semantic of `cp`, you should check that and if so copy the file to the dest dir without changing its name

Reply:
Did it so if we're downloading to an existing directory, then it's doing this.
Not sure how to handle something like that with a directory that doesn't exist _yet_, so say something like
```python
repo.download("src/file.py", "/non/existent/dir")
```

Not sure if it's worth trying to come up with a workable semantic there, because everyone will have their own weird edge cases. Hidden directories, filenames without extensions, dotfiles.
But then now we're behaving differently depending on whether the folder exists or not, so calling 
```python
repo.download("src/file.py", "/non/existent/dir")
```
on a machine where `/non/existent/dir` exists will save it to `/non/existent/dir/file.py`
and on a machine where it doesn't, it will save the file to `/non/existent/dir` (the file named "dir")

Also how would you expect the remote prefix to work here? Should it save to `dir/src/file.py` or `dir/file.py`?

Reply:
1. About missing directories - we should emulate the semantics of `cp -r` - so if you're attempting to download a file to a "non-existent-directory", that would mean that you should rename the file to the name of the non-existent directory. If you're downloading a directory, it should rename that directory to the non-existent directory. If you have a non-existent directory in the middle of the target path rather than its edge, then the whole copy command should throw an error (unless maybe a `mkdirs` flag is set to True)
2. If preserving the prefix, it means you assume that the target path is always a directory and you create your tree structure inside of it